### PR TITLE
Add support for starting the system from USB media

### PR DESCRIPTION
**Pull request by 150balbes**

### CODE REVIEW

File: include/configs/odroid-g12-common.h [hunk 1]
```diff
@@ -94,15 +94,16 @@
 #define ENV_PXE_DEFAULT
 #endif
 
-#define ENV_MMC_LIST_DEFAULT			"mmc_list=0 1\0"
+#define ENV_MMC_LIST_DEFAULT			"mmc_list=1 0\0"
```
Comment (tobetter) [inline include/configs/odroid-g12-common.h:95]: We've been requested to swap the boot order from SD to eMMC, but we'd like to keep them eMMC to SD by default. Instead, we are happy to provide simple instruction to swap them by a user in Linux userspace or Android using a tool and preparing to release them by the end of next week.

**Comment by 150balbes** [comment]
For proper operation of the startup with USB in files (boot.ini) and (boot.scr) you need to add this information. The new system is compatible with the old files that are used in the old images (the system will start normally with SD\eMMC).

https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2.ini#L98-L103

https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2-mainline.ini#L12-L17

File: include/configs/odroid-g12-common.h (hunk 1 above)
Comment (150balbes) [reply include/configs/odroid-g12-common.h:95]: To set the launch order, you can use a simple text file that will quickly change the launch order as desired by a specific user.

https://github.com/150balbes/Build-Armbian/blob/master/packages/bsp/odroid-n2/set_boot_order.ini

File: include/configs/odroid-g12-common.h [hunk 2]
```diff
@@ -126,6 +127,40 @@
...
+                "hdmimode=${hdmimode} modeline=${modeline} voutmode=${voutmode}; osd open; " \
+                "osd clear; vout output ${outputmode}; run set_spi_params; sf probe; " \
+                "sf read ${preloadaddr} ${start_kernel} ${size_kernel}; sf read ${initrd_high} " \
+                "${start_initrd} ${size_initrd}; sf read ${dtb_mem_addr} ${start_dtb} ${size_dtb}; " \
+                "if test -e mmc 1:1 spiboot.img; then fdt addr ${dtb_mem_addr}; fdt resize; " \
+                "fdt set /emmc@ffe07000 status 'disabled'; fdt set /soc/cbus/spifc@14000 " \
+                "status 'okay'; fi; bootm ${preloadaddr} ${initrd_high} ${dtb_mem_addr};" \
+                "\0"\
+        "boot_order=" \
+            "boot_usb boot_usb_distro boot_mmc boot_mmc_distro boot_rawimage boot_pxe boot_spi" \
```
Comment (tobetter) [inline include/configs/odroid-g12-common.h]: As described above regarding boot order of MMC cards, we also like to keep the current boot order, but willing to offer the capability to change it by the user.

File: include/configs/odroid-g12-common.h (hunk 1 above)
Comment (tobetter) [reply include/configs/odroid-g12-common.h:95]: > To set the launch order, you can use a simple text file that will quickly change the launch order as desired by a specific user.
> 
> https://github.com/150balbes/Build-Armbian/blob/master/packages/bsp/odroid-n2/set_boot_order.ini

I know, that's what I intended to add the value "mmc_list" but not improved yet. I've guided the similar instruction to many users through the forum or PM but some of them still having a difficulty to follow it.

Comment (150balbes) [reply include/configs/odroid-g12-common.h:95]: To use it, rename it (boot.ini) and write to any media (USB or SD card). By the way, I would recommend adding a mandatory shutdown of the system after the update procedure to the Piteboot update procedure. Now this procedure is repeated cyclically and you need to catch the moment and turn off N2.

**Review by tobetter — COMMENTED** [review]
Thank you for your PR, we are thinking of the boot order. But since we are getting different requests from different users, we prefer to keep the current booting order but as I described in the review, we are willing to provide the instruction to customize the boot order.

File: include/configs/odroid-g12-common.h (hunk 2 above)
Comment (150balbes) [reply include/configs/odroid-g12-common.h]: You are free to change this PR at your discretion. For me, it is important to be able to easily activate any desired user order in the SPI (by simply running a text file that is easy to edit to fit your needs).

Comment (tobetter) [reply include/configs/odroid-g12-common.h]: Thanks again, actually I am doing a similar thing to offer a method to change the boot order in U-boot but not have spent much time for some other work. And I am planning to keep doing so in Petitboot by mid of June, but I am feeling that I would provide some way what you did soon. :)

**Comment by tobetter** [comment]
> For proper operation of the startup with USB in files (boot.ini) and (boot.scr) you need to add this information. The new system is compatible with the old files that are used in the old images (the system will start normally with SD\eMMC).
> 
> https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2.ini#L98-L103
> 
> https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2-mainline.ini#L12-L17

Looks good to me, one suggestion in your change is to replace 'fatload' to 'load'. The command 'load' can manage the partition types of 'FAT32/64' and 'EXT2/3/4'.

**Comment by 150balbes** [comment]
That's reasonable, I'll change it in my files. By the way, I wouldn't spend a lot of time on Piteboot, u-boot-2019 for N2 with HDMI output support and direct systems intake control for startup should be out soon. :)

**Comment by tobetter** [comment]
> That's reasonable, I'll change it in my files. By the way, I wouldn't spend a lot of time on Piteboot, u-boot-2019 for N2 with HDMI output support and direct systems intake control for startup should be out soon. :)

Wow, that's awesome. Glad to hear the mainline U-boot can run on ODROID-N2 sometime soon. :)

**Comment by 150balbes** [comment]
Corrected version.

**Comment by 150balbes** [comment]
Hi  @tobetter . 
Are you planning to release a version of u-boot SPI (image) with USB support ?

**Comment by tobetter** [comment]
@150balbes No, I am not considering it at all.

**Pull request closed without merging**